crates/node/src/providers/robust_ecdsa.rs

```diff
@@ -0,0 +1,311 @@
+                }
+            },
+
+            _ => anyhow::bail!(
+                "eddsa task handler: received unexpected task id: {:?}",
+                channel.task_id()
+            ),
+        }
+        Ok(())
+    }
+
+    async fn spawn_background_tasks(self: Arc<Self>) -> anyhow::Result<()> {
+        let generate_presignatures = self
+            .per_domain_data
+            .iter()
+            .map(|(domain_id, data)| {
+                tracking::spawn(
+                    &format!("generate presignatures for domain {}", domain_id.0),
+                    presign::run_background_presignature_generation(
+                        self.client.clone(),
+                        self.mpc_config.clone(),
+                        self.config.presignature.clone().into(),
+                        *domain_id,
+                        data.presignature_store.clone(),
+                        data.keyshare.clone(),
+                    ),
+                )
+            })
+            .collect::<Vec<_>>();
+
+        for task in generate_presignatures {
+            task.await??;
+        }
+
+        Ok(())
+    }
+}
+
+/// Although currently the threshold is always equal to the number of signers, if in
+/// the future we might want to change that invariant, for example to achieve
+/// higher security guarantees for robust-ecdsa. In that case,
+/// this function enforces that the number of signers and the threshold
+/// computed below in `translate_threshold` stay consistent
+pub(super) fn get_number_of_signers(threshold: usize, _number_of_participants: usize) -> usize {
+    threshold
+}
+
+/// This function translates the current threshold from the contract
+/// to the threshold expected by the robust-ecdsa scheme, which
+/// is semantically different.
+/// The function should be no longer needed when these issues are solved:
+/// https://github.com/near/threshold-signatures/issues/255
+/// https://github.com/near/mpc/issues/1649
+pub(super) fn translate_threshold(
+    threshold: usize,
+    number_of_participants: usize,
+) -> anyhow::Result<usize> {
+    let number_of_signers = get_number_of_signers(threshold, number_of_participants);
+    anyhow::ensure!(number_of_signers >= 5, "Robust ECDSA requires the threshold to be at least 2, which implies that the number of signers to be at least 5");
+    Ok((number_of_signers - 1) / 2)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rstest::rstest;
+
+    // The resulting threshold for robust-ecdsa must always satisfy
+    // the underlying invariant that 2 * threshold + 1 <= number of signers
+    #[test]
+    fn test_translate_threshold() {
+        let max_size = 30;
+        for threshold in 5..max_size {
+            for number_of_participants in threshold..max_size {
+                let number_of_signers = get_number_of_signers(threshold, number_of_participants);
+                let new_threshold = translate_threshold(threshold, number_of_participants).unwrap();
```

Comment:
nit: can we add a check that `threshold >= (threshold-1)/2`?

Reply:
I don't really get this one. You mean `new_threshold >= (threshold-1)/2` ?